zetaclient/tss/crypto.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/pkg/errors"
+	"gitlab.com/thorchain/tss/go-tss/keysign"
+
+	"github.com/zeta-chain/node/pkg/chains"
+	"github.com/zeta-chain/node/pkg/cosmos"
+)
+
+// PubKey represents TSS public key in various formats.
+type PubKey struct {
+	cosmosPubKey cryptotypes.PubKey
+	ecdsaPubKey  *ecdsa.PublicKey
+}
+
+var (
+	base64Decode       = base64.StdEncoding.Decode
+	base64DecodeString = base64.StdEncoding.DecodeString
+	base64EncodeString = base64.StdEncoding.EncodeToString
+)
+
+// NewPubKeyFromBech32 creates a new PubKey from a bech32 address.
+// Example: `zetapub1addwnpepq2fdhcmfyv07s86djjca835l4f2n2ta0c7le6vnl508mseca2s9g6slj0gm`
+func NewPubKeyFromBech32(bech32 string) (PubKey, error) {
+	if bech32 == "" {
+		return PubKey{}, errors.New("empty bech32 address")
+	}
+
+	cosmosPubKey, err := cosmos.GetPubKeyFromBech32(cosmos.Bech32PubKeyTypeAccPub, bech32)
+	if err != nil {
+		return PubKey{}, errors.Wrap(err, "unable to GetPubKeyFromBech32")
+	}
+
+	pubKey, err := crypto.DecompressPubkey(cosmosPubKey.Bytes())
+	if err != nil {
+		return PubKey{}, errors.Wrap(err, "unable to DecompressPubkey")
+	}
+
+	return PubKey{
+		cosmosPubKey: cosmosPubKey,
+		ecdsaPubKey:  pubKey,
+	}, nil
+}
+
+// NewPubKeyFromECDSA creates a new PubKey from an ECDSA public key.
+func NewPubKeyFromECDSA(pk ecdsa.PublicKey) (PubKey, error) {
+	compressed := elliptic.MarshalCompressed(pk.Curve, pk.X, pk.Y)
+
+	return PubKey{
+		cosmosPubKey: &secp256k1.PubKey{Key: compressed},
+		ecdsaPubKey:  &pk,
+	}, nil
+}
+
+// NewPubKeyFromECDSAHexString creates PubKey from 0xABC12...
+func NewPubKeyFromECDSAHexString(raw string) (PubKey, error) {
+	if strings.HasPrefix(raw, "0x") {
+		raw = raw[2:]
+	}
+
+	b, err := hex.DecodeString(raw)
+	if err != nil {
+		return PubKey{}, errors.Wrap(err, "unable to decode hex string")
+	}
+
+	pk, err := crypto.UnmarshalPubkey(b)
+	if err != nil {
+		return PubKey{}, errors.Wrap(err, "unable to unmarshal pubkey")
+	}
+
+	return NewPubKeyFromECDSA(*pk)
+}
+
+// Bytes marshals pubKey to bytes either as compressed or uncompressed slice.
+//
+// In ECDSA, a compressed pubKey includes only the X and a parity bit for the Y,
+// allowing the full Y to be reconstructed using the elliptic curve equation,
+// thus reducing the key size while maintaining the ability to fully recover the pubKey.
+func (k PubKey) Bytes(compress bool) []byte {
+	pk := k.ecdsaPubKey
+	if compress {
+		return elliptic.MarshalCompressed(pk.Curve, pk.X, pk.Y)
+	}
+
+	return crypto.FromECDSAPub(pk)
+}
+
+// Bech32String returns the bech32 string of the public key.
+// Example: `zetapub1addwnpepq2fdhcmfyv07s86djjca835l4f2n2ta0c7le6vnl508mseca2s9g6slj0gm`
+func (k PubKey) Bech32String() string {
+	v, err := cosmos.Bech32ifyPubKey(cosmos.Bech32PubKeyTypeAccPub, k.cosmosPubKey)
+
+	// should not happen as we only set k.cosmosPubKey from the constructor
+	if err != nil {
+		panic("PubKey.Bech32String: " + err.Error())
```

Comment:
Why not still returning an error? It doesn't seem we have extensive usage of this function that would complicate the code.

Reply:
I insist on a simple API 😅  Cosmos also uses `MustMarhshal(...)` under the hood that theoretically might panic.

Made a safe workaround https://github.com/zeta-chain/node/pull/3170/commits/9b264d9d8816510374ee1b0f5d676021413baeef